src/net/neighbors/mod.rs

```diff
@@ -0,0 +1,430 @@
+                test_debug!(
+                    "{:?}: Throttle walk until {} to walk again (walk count: {}, walk retries: {})",
+                    &self.local_peer,
+                    self.walk_deadline,
+                    self.walk_count,
+                    self.walk_retries
+                );
+                return false;
+            }
+
+            // time to walk!
+            let new_walk = if let Some(w) = self.new_neighbor_walk(ibd) {
+                w
+            } else {
+                // unable to create a walk a this time.
+                return false;
+            };
+            self.walk = Some(new_walk);
+        }
+
+        return true;
+    }
+
+    #[cfg(test)]
+    fn print_walk_diagnostics(&mut self) {
+        let (mut inbound, mut outbound) = self.dump_peer_table();
+
+        inbound.sort();
+        outbound.sort();
+
+        debug!("{:?}: Walk finished ===================", &self.local_peer,);
+        debug!(
+            "{:?}: Peers outbound ({}): {}",
+            &self.local_peer,
+            outbound.len(),
+            outbound.join(", ")
+        );
+        debug!(
+            "{:?}: Peers inbound ({}):  {}",
+            &self.local_peer,
+            inbound.len(),
+            inbound.join(", ")
+        );
+
+        match PeerDB::get_frontier_size(self.peerdb.conn()) {
+            Ok(count) => {
+                debug!("{:?}: Frontier table size: {}", &self.local_peer, count);
+            }
+            Err(_) => {}
+        };
+        debug!("{:?}: Walk finished ===================", &self.local_peer);
+    }
+
+    #[cfg(not(test))]
+    fn print_walk_diagnostics(&self) {}
+
+    /// Update the state of our peer graph walk.
+    /// If we complete a walk, give back a walk result.
+    /// Mask errors by restarting the graph walk.
+    /// Returns the walk result, and a true/false flag to indicate whether or not the work for the
+    /// walk was finished (i.e. we either completed the walk, or we reset the walk)
+    pub fn walk_peer_graph(&mut self, ibd: bool) -> (bool, Option<NeighborWalkResult>) {
+        if !self.setup_walk(ibd) {
+            // nothing to do
+            return (true, None);
+        }
+
+        let mut walk = if let Some(walk) = self.walk.take() {
+            walk
+        } else {
+            // nothing to do
+            return (true, None);
+        };
+
+        let walk_result_opt = match walk.run(self) {
+            Ok(Some(ws)) => {
+                // walk ran to completion
+                self.walk_count += 1;
+                self.walk_deadline = self.connection_opts.walk_interval + get_epoch_time_secs();
+
+                debug!(
+                    "{:?}: walk has completed in {} steps ({} walks total)",
+                    &self.local_peer, walk.walk_step_count, self.walk_count
+                );
+
+                self.print_walk_diagnostics();
+                Some(ws)
```

Comment:
This branch returns `Some(ws)` if the walk ran to completion, and it is the only case in the match that continues execution of `walk_peer_graph`, but later on this function, the node:

1. checks if the walk timed out (which it shouldn't have if the walk completed, but also, if the walk completed before this later check, why should it time out now after the fact)
2. checks if there should be an occasional walk restart

If the walk timed out or should be restarted, this method returns `(true, None)` indicating that there's no more work to do on the walk. But if the walk completed, didn't time out, and didn't get the ocassional restart, this method returns `(false, Some(walk_result))`, indicating that work _does_ need to be done, even though the walk is finished. I guess I'm mostly confused about that last case: why is the boolean in that case `false`? Is there more work that needs to be done on the walk?

Reply:
Ah, I think the `Ok(None)` case for `walk.run()` should _not_ return.  We want the timeout check to apply on each pass as long as there's not an error.